### DISCUSSION TITLE
Extending context per workspace and type safety

### DISCUSSION THREAD

**Discussion by AlemTuzlak** [vitest-dev/vitest#3616] [Q&A] [answered]
Good day!

I have a question about vitest workspaces. I use vitest to run integration tests and unit tests which is a perfect use case for workspaces. 
In my setup file I do the following:
```jsx
beforeEach(ctx => {
  ctx.request = new Request("http://localhost");
  ctx.integration = integration;
});
```
and then I use a index.d.ts file to declare the types like this:
```ts
import * as integration from "./test-factory";
declare module "vitest" {
  export interface TestContext extends vitest.TestContext {
    request: Request;
    integration: typeof integration;
  }
}
```

and this allows me to use some builders in my integration tests, eg:
```ts
it("works", async ({ integration }) => {
   const user = await integration.createUser();
  expect(user).toBeTruthy()
})

```

Now my current issue is that if I accidentally try to use anything from the integration factory inside my UNIT tests it would try to connect to a db and fail and you wouldn't get a user back. Considering the type declarations are global I still can use the integration factory inside of my unit tests. 

If I split these two up into unit tests workspace and integration tests workspace is there a way for me to declare different contexts per workspace and have typescript infer the types and everything so I could maybe get the following typing in my unit tests:
```ts
beforeEach(ctx => {
  ctx.request = new Request("http://localhost");
  ctx.unit = unit;
});
```
```ts
import * as unit from "./unit-test-factory";
declare module "vitest" {
  export interface TestContext extends vitest.TestContext {
    request: Request;
    unit: typeof unit;
  }
}
```
and then have the following for my integration tests:
```ts
beforeEach(ctx => {
  ctx.request = new Request("http://localhost");
  ctx.integration = integration;
});
```
```ts
import * as integration from "./integration-test-factory";
declare module "vitest" {
  export interface TestContext extends vitest.TestContext {
    request: Request;
    integration: typeof integration;
  }
}
```
and then when I run unit tests this would complain:
```ts
// TS complains that integration doesn't exist in the unit test workspace
it("works", async ({ integration }) => {
   const user = await integration.createUser();
  expect(user).toBeTruthy()
})
```
and:
```ts
// TS complains that integration doesn't exist in the integration test workspace
it("works", async ({ unit }) => {
   const user = await unit.createUser();
  expect(user).toBeTruthy()
})
```

Thank you for the time and if this is not possible to do at the moment is there a way to make it work like this? I would look into contributing if yes!

Kind regards and happy coding!

**Comment by sheremet-va** [comment, marked as answer]
> Thank you for the time and if this is not possible to do at the moment is there a way to make it work like this? I would look into contributing if yes!

This is a TypeScript limitation, not Vitest limitation.

However, there is `test.extend` feature coming in the future version that should make contexts easier to work with: https://github.com/vitest-dev/vitest/pull/3554

**Reply by AlemTuzlak** [reply]
@sheremet-va Fully aware of that, I was just wondering if there is a way for vitest to infer the typing of the context based on the workspace it was initialized in. #3554 is interesting, but it does require me to import it everywhere as opposed to doing it globally in the setup file. Not sure how vitest works under the hood but it probably just runs my setup file before all the tests? It doesn't go through it and check if the context is extended?

**Reply by sheremet-va** [reply]
If your only concern is TypeScript, then this is the only good solution, because extending global in TypeScript extends everything globally (duh). 

You can try and play with putting different `tsconfig.json` files inside directories if you store tests in different folders, and include/exclude files that extend global vitest.

```
- unit
  - tsconfig.json: `{ "include": ["**/*.ts", "./setup-unit.ts"] }`
- integration
  - tsconfig.json: `{ "include": ["**/*.ts", "./setup-integration.ts"] }`
```

**Reply by AlemTuzlak** [reply]
@sheremet-va Alrighty, I just wanted to check if there is a way to do it in vitest without having to mess around with typescript. Alright I'll figure something out, another fight with typescript coming right up! Thank you so much for the input.
